### DISCUSSION TITLE
File Sharing Now Available

### DISCUSSION THREAD

**Discussion by hamza-usmani** [microsoft/WSA#321] [Announcements]
# File Sharing Now Available! (Build 2305)

## What is File Sharing?

We’ve been listening to the community, and many of you have been asking us to support sharing files between Windows and the Subsystem.  We are pleased to announce that Windows Subsystem for Android™ can now share your Windows user folders, like Documents and Pictures, with the Subsystem, so scenarios like uploading a photo to a social media app or editing a video in a creative app work seamlessly.  Folder sharing is enabled by default for preview users and can be toggled on and off in Windows Subsystem for Android™ Settings.  When folder sharing is enabled, your Windows user profile folder (e.g., “C:\Users\John Doe”) is shared as “/sdcard/Windows” in the Subsystem. 

## Privacy

Your privacy and security are important to us, so Android™ apps can only view or edit your files with your permission.  Apps must show a system dialog to request your permission, and permission can be revoked at any time from app settings.  The Subsystem also guards against malicious apps that would abuse your permission by scanning all Android™ apps with your antivirus software during installation and blocking threats, whether you installed the app from the Amazon Appstore or installed it with developer tools. 

## Limitations

There are a few restrictions you should be aware of: 

- Only files in your Windows user profile folder are available to the Subsystem.  Windows system folders, Program Files, other users’ folders, external drives, etc. are not supported. 
- Only files the Subsystem saves to “/sdcard/Windows” are available to Windows.  When opening files in other folders from an Android™ app, the Subsystem offers the option of copying the file to Windows and opening it there. 
- Windows hidden files/folders like AppData and Android hidden files/folders (dot prefixed names) are excluded from sharing. 
- Executable file types, like “.exe” are excluded from sharing to protect your security.  If the Subsystem tries to save a file with a blocked extension to “/sdcard/Windows”, the save will fail. 
- Cloud storage files must be downloaded to your device.  If your cloud storage provider includes online-only files in File Explorer, the online-only files will not be available to the Subsystem until you download them.  Simply open an online-only file in File Explorer to download it and make it available to the Subsystem.  You can also modify your cloud storage settings to make some or all of your files always available offline. 
- The Subsystem needs to index shared files before apps can access them.  Indexing is usually quick but may take longer if you have a lot of files.  Recent files are prioritized, so if you can’t find a file in an Android™ app, try opening it from File Explorer first. 

You can also share individual files with specific Android™ apps, even if folder sharing is disabled.  Simply drag and drop files from File Explorer to any open Android™ app that supports sharing files.  Some Android™ apps also support pasting files copied to the Windows clipboard.  In these cases, apps only receive read access to the shared file. 

We encourage you to try out file sharing scenarios and let us know how it works for you in Feedback Hub.  Thanks for using Windows Subsystem for Android™!

**Comment by HotCakeX** [comment]
This is amazing, thanks for the update 🙏🔥

This means the .apk files that I download on my computer are accessible in Android Subsystem and I can just install them without using ADB or any 3rd party tools 🤩

![image](https://github.com/microsoft/WSA/assets/118815227/03f2ceaa-7656-4f32-ac7d-026690256cf5)

**Reply by HotCakeX** [reply]
Although my OneDrive folder on my host:

![image](https://github.com/microsoft/WSA/assets/118815227/f0ef5742-82bc-496c-ac8f-9bdda661d950)

Doesn't appear in WSA. Some of the files are available offline but the rest are online-only. So is this a correct assumption that I'd need to make all of the files in OneDrive folder offline before the OneDrive folder itself can appear in WSA?

**Reply by jaholme** [reply]
That's correct. Android works by indexing files to make them available to apps, and if we indexed all online-only OneDrive files WSA would inadvertently cause all of them to be downloaded.

**Reply by HotCakeX** [reply]
@jaholme Thanks, there is no way to only index offline-only files and ignore the online-only files in OneDrive?
I have more than 300GB data on my OneDrive. I want to access ~100MB of them (that I marked as offline) in WSA. So in order to access those 100MB I need to mark the entire 300GB data as offline?

**Reply by kevinkieselbach** [reply]
The offline files in your OneDrive folder should be available in WSA as long as your OneDrive folder is located in your user profile folder.  Did you move your OneDrive folder to a secondary drive?  Otherwise, it may just take a while to index the files depending on how many files you have.  Try opening one of the files you're interested in from Windows File Explorer to add it to the Windows recent files list, which will cause WSA to index it immediately.  If you leave an app running in WSA, all the eligible files will eventually be indexed.

**Reply by HotCakeX** [reply]
## These are all of the folders that appear in Files app in WSA

![image](https://github.com/microsoft/WSA/assets/118815227/2555ff3b-a72c-46e6-a9ff-7fcb0d57e3ac)

![image](https://github.com/microsoft/WSA/assets/118815227/5e45258e-28e0-47b8-b773-d2727ec40063)

<br>

## These are the actual folders in my User directory

![image](https://github.com/microsoft/WSA/assets/118815227/7604482f-c1d1-43d2-87be-6df6ab10c330)

![image](https://github.com/microsoft/WSA/assets/118815227/928a423a-bb64-4c14-a3c5-3cf01853dc1c)

I didn't move the default OneDrive folder, there are many files in multiple folders already marked as offline but the OneDrive folder doesn't even appear in WSA.

in the past hour, at least one app was always open and I will continue to leave one open just in case.

I did open multiple pictures from my OneDrive as you suggested, no luck.

**Reply by HotCakeX** [reply]
This is my Windows indexing settings, i set it to Enhanced mode


![image](https://github.com/microsoft/WSA/assets/118815227/9ec737c5-6dbd-4a4a-9acf-547b1598504e)

**Reply by kevinkieselbach** [reply]
Your OneDrive folder is marked as a reparse point, and file sharing excludes reparse points, which explains your symptoms.  The goal was to exclude symbolic links and junctions, which would cause noisy access denied errors in Android, but your OneDrive folder is a different type of reparse point.  On my computer, my OneDrive for Business folder is a reparse point (but not a symlink/junction), but my personal OneDrive folder is not a reparse point.  We'll need to update the symlink/junction detection to be more precise.  In the meantime, if you copy the desired files to another folder in your user profile, WSA should find them.  We'll try to fix this within the next month.

**Reply by HotCakeX** [reply]
I see, thank you, glad to know that :-)

**Comment by subby2006** [comment]
Will the ability to browse Android files through the Windows Explorer be added anytime soon, similar to how you can browse Play (Android) files through the files app on ChromeOS?

**Reply by HotCakeX** [reply]
I just downloaded a file from a browser in WSA, copied it to Windows folder (using a file manager app i installed) and now I can access the same file in file explorer on my Windows 11 host. So, it's a 2-way access but only limited to a specific folder which is User folder.

**Reply by kevinkieselbach** [reply]
Our data indicates that most WSA users don't have any files in the Android /sdcard folders (Documents, Pictures, Music, Movies, Downloads, or DCIM), so making Windows files available to Android apps was the priority.  To transfer Android files to Windows, you can open the Android Files app from the WSA settings app and either 1) move/copy files to the "Windows" folder or 2) open the files and select the "Windows default preference" option to copy them to the Windows Downloads folder and open them on Windows.  Showing the Android Files contents directly in Windows File Explorer isn't on our roadmap at this time but feel free to make a feature request in Feedback Hub, which we'll consider if there's enough demand for it.

**Reply by jferdous70** [reply]
If a user has enabled WSA voluntarily, I believe they would benefit from the ability to see Android Files directly from Windows File Explorer. This would enable for seamless transfer of files and much easier to read.

**Comment by jaholme** [comment]
I misunderstood. Offline OneDrive files should be visible in WSA under Windows > OneDrive folder. If they're not showing up please file a Windows Feedback Hub report .

________________________________

**Reply by HotCakeX** [reply]
I just filed it, link: https://aka.ms/AAl1rc3

Prior to filing the feedback and adding a problem repo to it, I did change my OneDrive business to personal which has fewer files 25GB max cap and marked some files to be always available offline. then restart my computer and checked again, OneDrive still doesn't show up in WSA.

**Reply by kevinkieselbach** [reply]
Your OneDrive folder is marked as a reparse point, and file sharing excludes reparse points, which explains your symptoms. The goal was to exclude symbolic links and junctions, which would cause noisy access denied errors in Android, but your OneDrive folder is a different type of reparse point. On my computer, my OneDrive for Business folder is a reparse point (but not a symlink/junction), but my personal OneDrive folder is not a reparse point. We'll need to update the symlink/junction detection to be more precise. In the meantime, if you copy the desired files to another folder in your user profile, WSA should find them. We'll try to fix this within the next month.

**Comment by sapphired18** [comment]
my downloads and pictures is on other drives, i still want to access that, since my C:\ is small

**Reply by kevinkieselbach** [reply]
If your downloads/pictures folders are too large to fit on your C: drive, they would likely take a long time for Android to index.  You may be better off just copying the desired files into your user profile folder (such as to your desktop folder) when you want to access them from Android apps.  If this workaround isn't satisfactory, please file a feature request in Feedback Hub, which we'll consider if there's sufficient demand for it.

**Reply by ArthurKun21** [reply]
would it be possible to mount an external hdd drive in the future? my main drive is ssd that contains windows and it is quite small as I also store other dev programs

I was hoping to store most of my WSA files to my HDD drive.

**Reply by kevinkieselbach** [reply]
Next month's build (2306) will allow you to select a custom shared folder to use instead of the user profile folder.  WSA still won't include external drives by default since we don't know which drives are appropriate for Android to index as user data, but you could choose to share an external drive instead of your user profile folder with this change.

**Reply by HotCakeX** [reply]
Great news! Ty!

**Comment by sepehrsohrabi** [comment]
If we could access WSA files the same way WSL adds a Linux entry to the Explorer's sidebar, that would be amazing and very handy.
![Screenshot_20230603_124457_Chrome](https://github.com/microsoft/WSA/assets/31166217/bfd080d7-7ac5-427f-86f1-886e1a7ada66)

**Comment by Ulinsky** [comment]
Since it's now possible to drop .apk files into user folder, then go to android>files, and install those apk files normally from within WSA,
are there any plans to just let us install them from windows directly (basically what https://apps.microsoft.com/store/detail/wsatools-apk-installer-and-more/9N4P75DXL6FG?hl=de-at&gl=at&rtc=1 does)

**Comment by hgourvest** [comment]
Thank you very much for this update ! Unfortunately I can't get the drag'n drop function to work, can you share with us the technical details or the link in the Android documentation to which you referred to develop this feature?

**Reply by kevinkieselbach** [reply]
We implement the Android side of drag and drop by invoking an ACTION_SEND intent (https://developer.android.com/reference/android/content/Intent#ACTION_SEND) since most Android apps don't support drag and drop but many do support ACTION_SEND.

**Reply by HotCakeX** [reply]
It works in Telegram app, i tried.

**Comment by zhangz22** [comment]
Thanks for the update. Is there a reason that the entrypoint is chosen to be placed under `/storage/emulated/0/windows` instead of `/storage/SOME_FOLDER` (Makes the entrypoint to windows files looks like a SD Card)? 
A use case could be that when I want to use Android file cleaner apps to free up space from android storage, the cleaner would scan and include files from windows, making things unnecessary, slow and messy.

**Reply by kevinkieselbach** [reply]
We mount the Windows folder using a virtual file system, not a virtual block device, so it must be mounted under emulated storage rather than as a sdcard device for Android to be able to access it.  Emulated storage is also the primary storage, which is where apps look for files by default.  If we mounted the Windows folder outside of primary storage, many apps wouldn't find it since not all apps handle multiple storage volumes well.  We don't recommend "cleaner" apps on either Windows or Android.  They often delete things they shouldn't and cause system instability.

**Comment by jake-brown-maps** [comment]
Wanted to chime in and say nice work!

**Comment by TSchmider** [comment]
If the app, which is running in the subsystem, wants to save a file into the "/sdcard/Windows" directory, do it need to have the permission MANAGE_EXTERNAL_STORAGE?
I ask, because if I try to save a file, I get the error message "Operation not permitted".
Maybe you can provide me a code snippet. Thanks.

**Reply by kevinkieselbach** [reply]
To directly write to "/sdcard/Windows", you must have MANAGE_EXTERNAL_STORAGE permission, but the recommended best practice is to use Storage Access Framework's ACTION_CREATE_DOCUMENT to let the user choose where to save the file: https://developer.android.com/guide/topics/providers/document-provider.  This will open a folder picker showing the contents of /sdcard.  You can suggest that the user pick a subfolder of the "Windows" folder.

**Reply by TSchmider** [reply]
Thank you for your response. I will try your suggestion.

**Reply by TSchmider** [reply]
I tried your suggestion and it works - thank you for that.

Unfortunately I stumbled about another issue while testing. It seem that the Android file explorer
does not recognize any changes to the file size after its initial scan.

This can be verified using the Windows Subsystem for Android configuration application in Windows. 
1. Open the Android file explorer via System -> Files. 
2. Navigate to the Windows folder using the Android file explorer
3. Create a new file in this folder (e.g. text file) using the normal windows explorer.
4. Refresh the Android file explorer (now the file should have a size of 0 kb)
5. Insert text into this text file in Windows
6. Refresh the Android file explorer once again

The size of the file displayed in the android file explorer does not update.
![grafik](https://github.com/microsoft/WSA/assets/138693893/bc2d13f7-0b20-45bf-9614-68d07e0a2ded)
The size specifications are only updated after restarting the Windows Subsystem for Android service.
Are you already aware of this issue and if you are will this be fixed in any of the following versions?

**Reply by kevinkieselbach** [reply]
We're aware of that but are unlikely to fix it.  We currently subscribe to Windows file creation/deletion notifications and trigger Android media scans in response.  To fix this, we would also need to subscribe to Windows file modified notifications, which are substantially noisier.  I think the performance impact of constantly triggering media scans is prohibitive.  You should be able to access the file through the stream interfaces and just ignore the size column of the MediaStore database.

**Reply by TSchmider** [reply]
Thank you for your advise. I have tried it and it works.

**Comment by urazyjazzy** [comment]
Hi there, first of all thank you very much for finally enabling file sharing so that I can use the WSA like it is supposed to. But I have one small issue. I can create shortcuts on my taskbar for the apps I have installed but unable to create one for File app. Do I have to use a third party explorer app for this ? Second ; Is there a way to associate a filetype on Windows with a WSA app. For now I can not seem to open WSA apps automatically . They can not even bee seen as programs on Windows. So for a WSA app like Ireal Pro which only imports files via opening an html file  on another app  or via its forum page , 
So I have no other choice but save and share that file via whatever program I recieve the file from on the shared folder ,
 open WSA , 
open File app ,
 browse to the file you have shared 
and open it then let Android Subsytem do its thing and open the associated app ...
And additionally scrolling on a touchpad does not work like it supposed to. It keeps scrolling with some momentum instead of stopping
So do you guys have any suggestions or solutions for those potential bugs or problems ;)

**Comment by NiffirgkcaJ** [comment]
WSA won't access Windows folders no matter how much I turn on and off the feature, the fix that I have found so far is to reinstall the app. Is there any fix to this?

**Comment by stiller-leser** [comment]
I just tried this with an application that I installed via adb (as it is not available in the amazon app store). It say: "Ops! Unsupported source ddetected. Please install the app from the app store" - which I can't. Is there a way to fix this?

**Comment by urazyjazzy** [comment]
Hi there then it must be a problem within your windows version because I have two separate devices which have Windows 11 Pro on both with no problems. You could try to use a different android browser app and see if it helps...
Good Luck !!!

________________________________

**Comment by urazyjazzy** [comment]
Try the methods on this linkhttps://youtu.be/-8C1_VH-hvs?si=gcOmN1GSV_TH_PXlI can install any apk to my subsystem.If still you can try an emulator such as Bluestacks. Good luckUraz
-------- Original message --------From: Kaj-Sören Mossdorf ***@***.***> Date: 1/10/24  03:58  (GMT+03:00) To: microsoft/WSA ***@***.***> Cc: urazyjazzy ***@***.***>, Comment ***@***.***> Subject: Re: [microsoft/WSA] File Sharing Now Available (Discussion #321) 
I just tried this with an application that I installed via adb (as it is not available in the amazon app store). It say: "Ops! Unsupported source ddetected. Please install the app from the app store" - which I can't. Is there a way to fix this?

**Comment by urazyjazzy** [comment]
you have to enable developer mode under android subsystem settings then enable windows shared function if you already didn't do my friend. Good Luck
If that doesn't work I don't know what would anymore apart from disabling and resetting your subsystem...
Uraz
________________________________

**Comment by Dnis5** [comment]
Hello! I need support for something that may be simple, but for me, as a layman, it is difficult to resolve. I have files on the PC on the Local Disk E: drive that I need to transfer to the "Windows" folder of the Android Subsystem but I don't know how to do this. I read the entire article but I still don't know how to do it. If possible, I would like a step-by-step guide on how to do this or a tutorial link, that would be great. I appreciate all the help.
